lib/lfrfid/protocols/protocol_indala224.c

```diff
@@ -0,0 +1,348 @@
+#define INDALA224_ENCODED_DATA_LAST ((INDALA224_ENCODED_BIT_SIZE) / 8)
+
+#define INDALA224_DECODED_BIT_SIZE (224)
+#define INDALA224_DECODED_DATA_SIZE (28)
+
+#define INDALA224_US_PER_BIT (255)
+#define INDALA224_ENCODER_PULSES_PER_BIT (16)
+
+//----------------------------------------------------------------
+// Struct Definitions
+//----------------------------------------------------------------
+
+typedef struct {
+    uint8_t data_index;
+    uint8_t bit_clock_index;
+    bool last_bit;
+    bool current_polarity;
+    bool pulse_phase;
+} ProtocolIndala224Encoder;
+
+typedef struct {
+    uint8_t encoded_data[INDALA224_ENCODED_DATA_SIZE];
+    uint8_t negative_encoded_data[INDALA224_ENCODED_DATA_SIZE];
+    uint8_t corrupted_encoded_data[INDALA224_ENCODED_DATA_SIZE];
+    uint8_t corrupted_negative_encoded_data[INDALA224_ENCODED_DATA_SIZE];
+
+    uint8_t data[INDALA224_DECODED_DATA_SIZE];
+    ProtocolIndala224Encoder encoder;
+} ProtocolIndala224;
+
+//----------------------------------------------------------------
+// Utility Functions
+//----------------------------------------------------------------
+
+// Neater aliases for getting and setting bits
+
+bool gb(const uint8_t* data, size_t position) {
+    return bit_lib_get_bit(data, position);
+}
+
+void sb(uint8_t* data, size_t position, bool bit) {
+    bit_lib_set_bit(data, position, bit);
+}
+
+void psk1_to_psk2(uint8_t* data, size_t size) {
+    bool lastbit = gb(data, 0);
+    size_t num_bits = size * 8;
+
+    for(size_t i = 1; i < num_bits; i++) {
+        if(lastbit != gb(data, i)) {
+            lastbit = gb(data, i);
+            sb(data, i, 1);
+        } else {
+            sb(data, i, 0);
+        }
+    }
+}
+
+void psk2_to_psk1(uint8_t* data, size_t size) {
+    bool phase = 0;
+    size_t num_bits = size * 8;
+
+    for(size_t i = 0; i < num_bits; i++) {
+        if(gb(data, i) == 1) {
+            phase ^= 1;
+        }
+        sb(data, i, phase);
+    }
+}
+
+//----------------------------------------------------------------
+// Indala 224 Functions
+//----------------------------------------------------------------
+
+ProtocolIndala224* protocol_indala224_alloc(void) {
+    ProtocolIndala224* protocol = malloc(sizeof(ProtocolIndala224));
+    return protocol;
+};
+
+void protocol_indala224_free(ProtocolIndala224* protocol) {
+    free(protocol);
+};
+
+uint8_t* protocol_indala224_get_data(ProtocolIndala224* protocol) {
+    return protocol->data;
+};
+
+void protocol_indala224_decoder_start(ProtocolIndala224* protocol) {
+    memset(protocol->encoded_data, 0, INDALA224_ENCODED_DATA_SIZE);
+    memset(protocol->negative_encoded_data, 0, INDALA224_ENCODED_DATA_SIZE);
+    memset(protocol->corrupted_encoded_data, 0, INDALA224_ENCODED_DATA_SIZE);
+    memset(protocol->corrupted_negative_encoded_data, 0, INDALA224_ENCODED_DATA_SIZE);
+};
+
+static bool protocol_indala224_check_preamble(uint8_t* data, size_t bit_index) {
```

Comment:
If you change this function to this:

`static bool protocol_indala224_check_preamble(uint8_t* data, size_t bit_index) {`
`    // Preamble 10000000__00000000__00000000__00000`
`    if(bit_lib_get_bits_32(data, bit_index, 30) ==`
`       0b100000000000000000000000000000) {`
`        return true;`
`    } else {`
`        return false;`
`    }`
`}`

This entire thing will work.

Reply:
Can you help me bro for My indala card ?